image_segmentation/pytorch/data_loading/data_loader/unet3d_data_loader.py

```diff
@@ -0,0 +1,139 @@
+    assert len(data) > 0, f"Found no data at {path}"
+    return data
+
+
+def load_data(path, files_pattern):
+    data = sorted(glob.glob(os.path.join(path, files_pattern)))
+    assert len(data) > 0, f"Found no data at {path}"
+    return data
+
+
+def get_split(data, train_idx, val_idx):
+    train = list(np.array(data)[train_idx])
+    val = list(np.array(data)[val_idx])
+    return train, val
+
+
+def split_eval_data(x_val, y_val, num_shards, shard_id):
+    x = [a.tolist() for a in np.array_split(x_val, num_shards)]
+    y = [a.tolist() for a in np.array_split(y_val, num_shards)]
+    return x[shard_id], y[shard_id]
+
+
+def get_data_split(path: str, num_shards: int, shard_id: int):
+    with open("evaluation_cases.txt", "r") as f:
+        val_cases_list = f.readlines()
+    val_cases_list = [case.rstrip("\n") for case in val_cases_list]
+    imgs = load_data(path, "*_x.npy")
+    lbls = load_data(path, "*_y.npy")
+    assert len(imgs) == len(lbls), f"Found {len(imgs)} volumes but {len(lbls)} corresponding masks"
+    imgs_train, lbls_train, imgs_val, lbls_val = [], [], [], []
+    for (case_img, case_lbl) in zip(imgs, lbls):
+        if case_img.split("_")[-2] in val_cases_list:
+            imgs_val.append(case_img)
+            lbls_val.append(case_lbl)
+        else:
+            imgs_train.append(case_img)
+            lbls_train.append(case_lbl)
+    mllog_event(key="train_samples", value=len(imgs_train), sync=False)
+    mllog_event(key="eval_samples", value=len(imgs_val), sync=False)
+    imgs_val, lbls_val = split_eval_data(imgs_val, lbls_val, num_shards, shard_id)
+    return imgs_train, imgs_val, lbls_train, lbls_val
+
+
+class SyntheticDataset(Dataset):
+    def __init__(
+        self,
+        channels_in=1,
+        channels_out=3,
+        shape=(128, 128, 128),
+        device="cpu",
+        layout="NCDHW",
+        scalar=False,
+    ):
+        shape = tuple(shape)
+        x_shape = (channels_in,) + shape if layout == "NCDHW" else shape + (channels_in,)
+        self.x = torch.rand(
+            (32, *x_shape), dtype=torch.float32, device=device, requires_grad=False
+        )
+        if scalar:
+            self.y = torch.randint(
+                low=0,
+                high=channels_out - 1,
+                size=(32, *shape),
+                dtype=torch.int32,
+                device=device,
+                requires_grad=False,
+            )
+            self.y = torch.unsqueeze(self.y, dim=1 if layout == "NCDHW" else -1)
+        else:
+            y_shape = (channels_out,) + shape if layout == "NCDHW" else shape + (channels_out,)
+            self.y = torch.rand(
+                (32, *y_shape), dtype=torch.float32, device=device, requires_grad=False
+            )
+
+    def __len__(self):
+        return 64
+
+    def __getitem__(self, idx):
+        return self.x[idx % 32], self.y[idx % 32]
+
+
+def get_data_loaders(flags: Namespace, num_shards: int, global_rank: int, device: torch.device):
```

Comment:
Looks like we need a default value for `global_rank` - i.e. when `torch_xla == False`.

Reply:
Could you please clarify why a default value for `global_rank` is needed?

When we are using native PyTorch (i.e. when `torch_xla == False`), we do not need to pass `global_rank` to `cl.get_data_loaders` to construct the data loaders. The `torch.utils.data.distributed.DistributedSampler` object will automatically discover the global rank of the current process from its pytorch distributed group.

The `global_rank` argument is only used for `get_data_split` and for constructing `pl.MpDeviceLoader` in `xl.get_data_loaders`. Since each distributed process should have a different `global_rank`, I don't see why we should give a default value to `global_rank`?

Reply:
Right. Since you only need this variable for `torch_xla == True`, the user should be free to NOT pass a value for `global_rank` when `torch_xla == False`. What am I missing?

Reply:
Are you saying when `torch_xla == False`, `global_rank=local_rank` is passed?

Reply:
Oh I see the confusion. To clarify, for `torch_xla == False`, the user must also pass in `global_rank`. This is because the call to [`get_data_split`](https://github.com/thisisalbertliang/training/blob/925ddc9e2bcd221ef596b924d1aaa067d6c7b662/image_segmentation/pytorch/data_loading/data_loader/unet3d_data_loader.py#L118-L120) (which is for splitting the **validation** dataset for data parallelism) requires the `global_rank` argument.

So I believe regardless of whether `torch_xla` is `False` or `True`, `global_rank` should always be a required argument.